### ISSUE TITLE
DocumentDB Parameter Group shows constant changes to profiler_sampling_rate

### ISSUE THREAD

**Issue by AaronKalair** [hashicorp/terraform-provider-aws#21679]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version
```
terraform -v
Terraform v1.0.8
on darwin_amd64
+ provider registry.terraform.io/hashicorp/archive v2.2.0
+ provider registry.terraform.io/hashicorp/aws v3.61.0
```

### Affected Resource(s)
* aws_docdb_cluster_parameter_group - https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/docdb_cluster_parameter_group

### Terraform Configuration Files
```hcl
locals {
  parameters = {
    audit_logs : "enabled"
    change_stream_log_retention_duration : 10800 
    profiler : "enabled"  
    profiler_sampling_rate : 1 
    profiler_threshold_ms : 100 
    tls : "enabled" 
    ttl_monitor : "enabled"
  }
}

resource "aws_docdb_cluster_parameter_group" "parameter_group" {
  family      = "docdb3.6"
  name        = "shared-parameter-group"
  description = "Shared by most clusters. Any changes will affect all clusters using it."

  dynamic "parameter" {
    for_each = local.parameters

    content {
      name  = parameter.key
      value = parameter.value
    }
  }
  tags = {
    Environment   = "Staging"
    ProvisionedBy = "terraform"
  }
}
```

### Expected Behavior
* After the DocumentDB ParameterGroup has been created, unless the value of `profiler_sampling_rate` changes Terraform Plans show no diffs

### Actual Behavior
* Every Terraform Plan shows this

```
      + parameter {
          + apply_method = "pending-reboot"
          + name         = "profiler_sampling_rate"
          + value        = "1"
        }
      - parameter {
          - apply_method = "pending-reboot" -> null
          - name         = "profiler_sampling_rate" -> null
          - value        = "1.0" -> null
        }
```

### Steps to Reproduce

1. Create a DocumentDB Parameter group using the above code
2. Apply the changes
3. Run a new plan
4. Notice that despite no values changing Terraform claims the profiler_sampling_rate parameter has changed

### References
* https://github.com/hashicorp/terraform-provider-aws/issues/13384

**Comment by justinretzolk** [comment]
Hey @AaronKalair 👋 Thank you for taking the time to file this issue. I noticed that the diff appears to be changing the `parameter.value` from `1.0` to `1` (the `1` being what's defined in your `local.parameters.profiler_sampling_rate` value). With that in mind,I took a look over the [DocumentDB Cluster Parameters Reference](https://docs.aws.amazon.com/documentdb/latest/developerguide/cluster_parameter_groups-list_of_parameters.html), and it looks like the value for `profiler_sampling_rate` is expected to be a `float`, which I believe is what's causing the perpetual diff here (I imagine AWS is converting the `int` to a `float` on their end, but Terraform will see this as a difference when it's read back). 

Can you try setting the value to a `float` rather than an `int`, and see if that corrects the issue?

**Comment by AaronKalair** [comment]
Hey,

Thanks for the response, even when we explicitly set it to a float e.g.

`profiler_sampling_rate : 1.0`

It still causes a constant diff in our Terraform plans that looks like:

```
      + parameter {
          + apply_method = "pending-reboot"
          + name         = "profiler_sampling_rate"
          + value        = "1"
        }
      - parameter {
          - apply_method = "pending-reboot" -> null
          - name         = "profiler_sampling_rate" -> null
          - value        = "1.0" -> null
        }
```

Bizarrely it looks like even though we set a float it ends up with `value = 1` 

I can't seem to find a way to force a cast to float


Interestingly though I set the value to `1.1` and ran a `terraform plan`

```
      + parameter {
          + apply_method = "pending-reboot"
          + name         = "profiler_sampling_rate"
          + value        = "1.1"
        }
```

And it looks like the float value is kept there, is this some bizarre quirk of Terraform assuming 1.0 and 1 are identical and converting them somewhere?

**Comment by justinretzolk** [comment]
Thanks for the update @AaronKalair - I did a bit of digging with that in mind and wanted to report back with what I've found. Unfortunately, I didn't make it all the way down the rabbit hole to find the _exact_ cause, but I have some suspicions.

The first thing I noticed was that the `value` argument [is a `TypeString`](https://github.com/hashicorp/terraform-provider-aws/blob/ca2cb2838b07e373baa7ca5bbc5d0778509b493d/internal/service/docdb/cluster_parameter_group.go#L74-L77), more on this further down.

In order to process the parameters, the resource uses the [`expandParameters` function](https://github.com/hashicorp/terraform-provider-aws/blob/ca2cb2838b07e373baa7ca5bbc5d0778509b493d/internal/service/docdb/flex.go#L8-L28). This function gets the string value of `value` using type assertion (`data["value"].(string)`) and then also uses `aws.String()` to get back a pointer to the string value.

I suspect that somewhere along the way (this is what I didn't quite fully figure out yet), the `int`/`float` that you're passing is being converted to a `string`, and that wherever that's happening, `1.0` is being converted to `"1"`, while `1.1` is being converted (appropriately, in this case) to `"1.1"`. I don't believe that this is happening at the `expandParameters` function, as I wouldn't expect the type assertion to work in that case. That leads me to believe it's happening prior to that, perhaps at the Terraform SDK level, but I've not had enough time to fully track down _where_ that would be happening.

Someone else with a stronger Go background may be able to determine _exactly_ what's happening here, but in the meantime, I suspect that updating the `profiler_sampling_rate` in the `locals` block to the following might unblock you. If this doesn't work, I'll mark this as a bug so that we can continue to look into it further. 

```hcl
profiler_sampling_rate : "1.0"
```

**Comment by AaronKalair** [comment]
Brilliant!

I just tried this and setting it to "1.0" has fixed the non stop diffs.

Thank you for your help, is it worth me noting this on the documentation page for DocDB Parameter Group?

**Comment by justinretzolk** [comment]
I'm glad that solved it @AaronKalair! I definitely see the merits in updating the documentation - I'm happy to do so, so as to not take up more of your time. Since you ran into this, would you mind if I tag you in the pull request so that you can see my proposed change? I'd love your perspective, given that you ran into this, if you don't mind.

**Comment by AaronKalair** [comment]
Yeah that sounds good happy to review!
